tests/integration/cmd_link_unlink_test.go

```diff
@@ -9,138 +13,201 @@ import (
-				helper.Cmd("odo", "url", "create", "--port", "8080", "--context", backendContext).ShouldPass()
-				helper.Cmd("odo", "push", "--context", backendContext).ShouldPass()
-
-				// link both component and wait till frontend dc rollout properly
-				helper.Cmd("odo", "link", "backend", "--port", "8080", "--wait", "--context", frontendContext).ShouldPass()
-				helper.HttpWaitFor(frontendURL, "Hello world from node.js!", 20, 1)
-
-				// ensure that the proper envFrom entry was created
-				envFromOutput := oc.GetEnvFromEntry("frontend", appName, commonVar.Project)
-				Expect(envFromOutput).To(ContainSubstring("backend"))
-
-				helper.Cmd("odo", "unlink", "backend", "--port", "8080", "--context", frontendContext).ShouldPass()
+			When("the link is pushed", func() {
+				JustBeforeEach(func() {
+					helper.Cmd("odo", "push", "--context", frontendContext).ShouldPass()
+				})
+				It("should ensure that the proper envFrom entry was created", func() {
+					envFromOutput := runner.GetEnvFromEntry(frontendComp, "app", commonVar.Project)
+					Expect(envFromOutput).To(ContainSubstring(backendComp))
+					helper.HttpWaitFor(frontendURL, "Hello world from node.js!", 20, 1)
+				})
+				It("should find the link and environment variables in odo describe", func() {
+					checkDescribe(frontendContext, backendComp, true, false)
+				})
+				It("should find the linked environment variable", func() {
+					stdOut := helper.Cmd("odo", "exec", "--context", frontendContext, "--", "sh", "-c", "echo $SERVICE_BACKEND_IP").ShouldPass().Out()
+					Expect(stdOut).To(Not(BeEmpty()))
+				})
+				It("should not allow re-linking", func() {
+					outputErr := helper.Cmd("odo", "link", backendComp, "--context", frontendContext).ShouldFail().Err()
+					Expect(outputErr).To(ContainSubstring("already linked"))
+				})
+
+				It("should successfully delete component after linked component is deleted", func() {
+					// Testing: https://github.com/openshift/odo/issues/2355
+					helper.Cmd("odo", "delete", "-f", "--context", backendContext).ShouldPass()
+					helper.Cmd("odo", "delete", "-f", "--context", frontendContext).ShouldPass()
+				})
+
+				When("unlinking the two components", func() {
+					JustBeforeEach(func() {
+						helper.Cmd("odo", "unlink", backendComp, "--context", frontendContext).ShouldPass()
+					})
+					It("should find the link in odo describe", func() {
+						checkDescribe(frontendContext, backendComp, true, false)
+					})
+					It("should not allow unlinking again", func() {
+						stdOut := helper.Cmd("odo", "unlink", backendComp, "--context", frontendContext).ShouldFail().Err()
+						Expect(stdOut).To(ContainSubstring(fmt.Sprintf("failed to unlink the component %q since no link was found in the configuration referring this component", backendComp)))
+					})
+
+					When("odo push is executed", func() {
```

Comment:
If we have only one test spec. we should avoid using nested when, no?

Reply:
But what if we encounter a case where we will need to test something after the push is executed, hence the nesting.

Reply:
> But what if we encounter a case where we will need to test something after the push is executed, hence the nesting.

IMO if we get such scenario in future, we will hit this nesting thing. Meanwhile I would suggest to avoid nesting as much as possible.

Reply:
Well, I just realized I missed a case there, see https://github.com/openshift/odo/pull/4891/commits/5df7d7b152079df6aa2a0ecf1ed0567b5778c5da.